File: RecordDatabaseNative/lib/dialogs.cpp
#include "dialogs.h"

#include <QtWidgets>

QRegExpValidator validatorId(QRegExp("^\\d{6,}$"));

NewRecordDialog::NewRecordDialog(QWidget* parent) : QDialog(parent)
{
    auto idLabel = new QLabel(tr("Record ID:"));
    m_idLineEdit = new QLineEdit;
    m_idLineEdit->setValidator(&validatorId);

    auto nameLabel = new QLabel(tr("Name:"));
    m_nameLineEdit = new QLineEdit;

    auto surnameLabel = new QLabel(tr("Surname:"));
    m_surnameLineEdit = new QLineEdit;

    auto formLayout = new QFormLayout;
    formLayout->addRow(idLabel, m_idLineEdit);
    formLayout->addRow(nameLabel, m_nameLineEdit);
    formLayout->addRow(surnameLabel, m_surnameLineEdit);

    auto createRecordButton = new QPushButton(tr("Create Record"));
    createRecordButton->setEnabled(false);

    auto buttonBox = new QDialogButtonBox(Qt::Horizontal);
    buttonBox->addButton(createRecordButton, QDialogButtonBox::AcceptRole);
    buttonBox->setStandardButtons(QDialogButtonBox::Cancel);

    auto layout = new QVBoxLayout();
    layout->addLayout(formLayout);
    layout->addWidget(buttonBox);

    setLayout(layout);

    // CONNECTIONS
    connect(m_idLineEdit, &QLineEdit::textChanged, this, [this, createRecordButton](){
        createRecordButton->setEnabled(m_idLineEdit->hasAcceptableInput());
    });
    connect(m_nameLineEdit, &QLineEdit::textChanged, this, [this](const QString& text){
        m_nameLineEdit->setText(QLocale().toUpper(text));
    });
    connect(m_surnameLineEdit, &QLineEdit::textChanged, this, [this](const QString& text){
        m_surnameLineEdit->setText(QLocale().toUpper(text));
    });

    connect(buttonBox, &QDialogButtonBox::accepted, this, &NewRecordDialog::beforeAccept);
    connect(buttonBox, &QDialogButtonBox::rejected, this, &NewRecordDialog::reject);
}

void NewRecordDialog::open()
{
    m_idLineEdit->clear();
    m_nameLineEdit->clear();
    m_surnameLineEdit->clear();

    QDialog::open();
}

void NewRecordDialog::beforeAccept()
{
    auto recordId = m_idLineEdit->text();
    auto name = m_nameLineEdit->text();
    auto surname = m_surnameLineEdit->text();

    emit recordCreated(recordId, name, surname);
}

File: RecordDatabasePython/recorddatabase/dialogs.py
from PyQt5.QtCore import *
from PyQt5.QtWidgets import *
from PyQt5.QtGui import QRegExpValidator


validatorId = QRegExpValidator(QRegExp("^\\d{6,}$"))


class NewRecordDialog(QDialog):
    recordCreated = pyqtSignal(str, str, str)

    def __init__(self, parent):
        super().__init__(parent)

        idLabel = QLabel(self.tr("Record ID:"))
        idLineEdit = QLineEdit()
        idLineEdit.setValidator(validatorId)

        nameLabel = QLabel(self.tr("Name:"))
        nameLineEdit = QLineEdit()

        surnameLabel = QLabel(self.tr("Surname:"))
        surnameLineEdit = QLineEdit()

        formLayout = QFormLayout()
        formLayout.addRow(idLabel, idLineEdit)
        formLayout.addRow(nameLabel, nameLineEdit)
        formLayout.addRow(surnameLabel, surnameLineEdit)

        createRecordButton = QPushButton(self.tr("Create Record"))
        createRecordButton.setEnabled(False)

        buttonBox = QDialogButtonBox(Qt.Horizontal)
        buttonBox.addButton(createRecordButton, QDialogButtonBox.AcceptRole)
        buttonBox.setStandardButtons(QDialogButtonBox.Cancel)

        layout = QVBoxLayout()
        layout.addLayout(formLayout)
        layout.addWidget(buttonBox)

        self.setLayout(layout)

        # CONNECTIONS
        idLineEdit.textChanged.connect(lambda: createRecordButton.setEnabled(idLineEdit.hasAcceptableInput()))
        nameLineEdit.textEdited.connect(lambda text: nameLineEdit.setText(QLocale().toUpper(text)))
        surnameLineEdit.textEdited.connect(lambda text: surnameLineEdit.setText(QLocale().toUpper(text)))

        buttonBox.accepted.connect(self.beforeAccept)
        buttonBox.rejected.connect(self.reject)

        # PROPERTIES
        self.idLineEdit = idLineEdit
        self.nameLineEdit = nameLineEdit
        self.surnameLineEdit = surnameLineEdit

    def beforeAccept(self):
        recordId = self.idLineEdit.text()
        name = self.nameLineEdit.text()
        surname = self.surnameLineEdit.text()

        self.recordCreated.emit(recordId, name, surname)

    def open(self):
        self.idLineEdit.clear()
        self.nameLineEdit.clear()
        self.surnameLineEdit.clear()

        super().open()


class OpenRecordDialog(QDialog):
    recordSelected = pyqtSignal(str)

    def __init__(self, parent):
        super().__init__(parent)

        # DIALOG DEFINITION
        idLabel = QLabel(self.tr("Record ID:"))
        idLineEdit = QLineEdit()
        idLineEdit.setValidator(validatorId)

        openButton = QPushButton(self.tr("Open"))

        formLayout = QFormLayout()
        formLayout.addRow(idLabel, idLineEdit)

        layout = QVBoxLayout()
        layout.addLayout(formLayout)
        layout.addWidget(openButton)

        self.setLayout(layout)

        # CONNECTIONS
        openButton.clicked.connect(lambda: self.recordSelected.emit(idLineEdit.text()))

        # PROPERTIES
        self.idLineEdit = idLineEdit

    def open(self):
        self.idLineEdit.clear()

        super().open()
